src/hotspot/share/opto/library_call.cpp

```diff
@@ -2828,10 +2829,35 @@ bool LibraryCallKit::inline_unsafe_allocate() {
     test = _gvn.transform(new SubINode(inst, bits));
     // The 'test' is non-zero if we need to take a slow path.
   }
-
   Node* obj = new_instance(kls, test);
+#if INCLUDE_JVMTI
+  // Check if JvmtiExport::_should_post_allocation_notifications is enabled and post notifications
+  IdealKit ideal(this);
+  IdealVariable result(ideal); ideal.declarations_done();
+  Node* ONE = ideal.ConI(1);
+  Node* addr = makecon(TypeRawPtr::make((address) &JvmtiExport::_should_post_allocation_notifications));
+  Node* should_post_vm_object_alloc = ideal.load(ideal.ctrl(), addr, TypeInt::BOOL, T_BOOLEAN, Compile::AliasIdxRaw);
+
+  ideal.sync_kit(this);
+  ideal.if_then(should_post_vm_object_alloc, BoolTest::eq, ONE); {
+    const TypeFunc *tf = OptoRuntime::notify_allocation_Type();
+    address funcAddr = OptoRuntime::notify_allocation();
+    sync_kit(ideal);
+    Node* call = make_runtime_call(RC_NO_LEAF, tf, funcAddr, "_notify_allocation", TypePtr::BOTTOM, obj);
+    ideal.sync_kit(this);
+    ideal.set(result,_gvn.transform(new ProjNode(call, TypeFunc::Parms+0)));
+  } ideal.else_(); {
+    ideal.set(result,obj);
+  } ideal.end_if();
+  final_sync(ideal);
+
+  set_result(ideal.value(result));
+  return true;
+#else
```

Comment:
Nit: It is better to replace #else at 2856 with #endif. Then #endif at 2859 is not needed.

Reply:
I wonder if it possible to do the following:
```
+   . . . . . . .
+   final_sync(ideal);
+   obj = ideal.value(result);
+   return true;
+ #endif // INCLUDE_JVMTI
  set_result(obj);
  return true;
```
But I'm not sure if one more sync is needed in such a case.
At lease, this line should not be under #if/#else:
`  return true;`